src/Sarif/Writers/SarifConsolidator.cs

```diff
@@ -0,0 +1,259 @@
+                }
+            }
+        }
+
+        public void Consolidate(ThreadFlow tFlow)
+        {
+            if (tFlow.Locations != null)
+            {
+                TotalThreadFlowLocations += tFlow.Locations.Count;
+
+                for (int i = 0; i < tFlow.Locations.Count; ++i)
+                {
+                    ThreadFlowLocation tfl = tFlow.Locations[i];
+                    Trim(tfl.Location);
+
+                    if (tfl.Index == -1)
+                    {
+                        if (!_uniqueThreadFlows.TryGetValue(tfl, out int tflIndex))
+                        {
+                            tflIndex = _run.ThreadFlowLocations.Count;
+                            _run.ThreadFlowLocations.Add(tfl);
+                            _uniqueThreadFlows[tfl] = tflIndex;
+                        }
+
+                        tFlow.Locations[i] = new ThreadFlowLocation() { Index = tflIndex };
+                    }
+                }
+            }
+        }
+
+        public IList<Location> Trim(IList<Location> locations)
+        {
+            if (locations == null || locations.Count == 0) { return locations; }
+
+            HashSet<Location> uniqueLocations = new HashSet<Location>(Location.ValueComparer);
+            List<Location> newLocations = new List<Location>();
+
+            this.TotalLocations += locations.Count;
+            for (int i = 0; i < locations.Count; ++i)
+            {
+                Location loc = locations[i];
+
+                // Trim Locations themselves
+                Trim(loc);
+
+                // Keep only one copy of each unique location
+                if (uniqueLocations.Add(loc))
+                {
+                    this.UniqueLocations++;
+                    newLocations.Add(loc);
+                }
+            }
+
+            return newLocations;
+        }
+
+        public void Trim(Location loc)
+        {
+            if (loc != null)
+            {
+                // Remove Ids
+                loc.Id = -1;
+
+                Trim(loc.PhysicalLocation);
+                Trim(loc.LogicalLocation);
+                Trim(loc.LogicalLocations);
+            }
+        }
+
+        public IList<LogicalLocation> Trim(IList<LogicalLocation> locations)
+        {
+            if (locations == null || locations.Count == 0) { return locations; }
+
+            HashSet<LogicalLocation> uniqueLocations = new HashSet<LogicalLocation>(LogicalLocation.ValueComparer);
+            List<LogicalLocation> newLocations = new List<LogicalLocation>();
+
+            this.TotalLocations += locations.Count;
+            for (int i = 0; i < locations.Count; ++i)
+            {
+                LogicalLocation loc = locations[i];
+
+                // Trim Locations themselves
+                Trim(loc);
+
+                // Keep only one copy of each unique location
+                if (uniqueLocations.Add(loc))
+                {
+                    this.UniqueLocations++;
```

Comment:
>this.UniqueLocations++; [](start = 20, length = 23)

you could keep incrementing this, or you could simply compute this value a single time before return (set the property to uniqueLocations.Count)

Reply:
Good idea, fixed.